Microsoft.Azure.Cosmos/src/Tracing/Trace.cs

```
@@ -122,20 +123,38 @@ public static Trace GetRootTrace(
                 summary: new TraceSummary());
         }
 
+        /// <summary>
+        /// Adds a datum to the this trace instance.
+        /// This method is thread-safe.
+        /// </summary>
+        /// <param name="key">The key to associate the datum.</param>
+        /// <param name="traceDatum">The datum itself.</param>
         public void AddDatum(string key, TraceDatum traceDatum)
         {
-            this.data.Value.Add(key, traceDatum);
+            this.data.Value.TryAdd(key, traceDatum);
             this.Summary.UpdateRegionContacted(traceDatum);
         }
 
+        /// <summary>
+        /// Adds a datum to the this trace instance.
+        /// This method is thread-safe.
+        /// </summary>
+        /// <param name="key">The key to associate the datum.</param>
+        /// <param name="value">The datum itself.</param>
         public void AddDatum(string key, object value)
         {
-            this.data.Value.Add(key, value);
+            this.data.Value.TryAdd(key, value);
```

Same here - This code change modifies the behavior in a possibly breaking manner. Dictionary.Add throws an ArgumentException when the key already exists. ConcurrentDictionary.TryAdd will silently ignore that fact. I think it would be safer to maintaint that API contract - so, if TryAdd retruns false it should thrown an ArgumentException simulating what Dictionary.Add is doing.